frontend/src/pages/main/home/client/components/graphic/Graphic.tsx

```diff
@@ -0,0 +1,117 @@
+import {Content} from "../../../../../../common/constansts/content";
+import {Transaction, TransactionType} from "../../../../../../types/transactions";
+import {getTransactionForYearService} from "../../../../../../services/transaction";
+import {useAppSelector} from "../../../../../../stores/redux/hooks";
+import {selectLoggedUser} from "../../../../../../stores/redux/loggedUser/loggedUserSlice";
+import {CURRENCY_USD} from "../../../../../../common/constansts/currency";
+import {OTHER_TRANSACTIONS, THRESHOLD_UNIQUE_TRANSACTIONS} from "../../../../../../common/constansts/transactions";
+import {COLORS} from "../../../../../../common/constansts/colors";
+import Finance from "../finance/Finance";
+
+Chart.register(Tooltip, Legend, ArcElement);
+
+const initialTransactions: Transaction[] = [];
+
+const Graphic: React.FC = () => {
+
+    const loggedUser = useAppSelector(selectLoggedUser);
+    const [transactions, setTransactions] = useState<Transaction[]>(initialTransactions);
+    const [income, setIncome] = useState<number>(0);
+    const [expenses, setExpenses] = useState<number>(0);
+
+    const updateTransactions = async () => {
+        const transactionData = await getTransactionForYearService(loggedUser.id, loggedUser.userType, new Date().getFullYear(), new Date().getFullYear());
+        const updatedTransactions = mergeTransactions(transactionData);
+        setTransactions(updatedTransactions);
+        const expenses = updatedTransactions.reduce((total, transaction) => total + transaction.amount, 0);
+        setExpenses(expenses);
+        const income = transactionData.filter(transaction => transaction.TransactionType === TransactionType.TOP_UP)
+                                               .reduce((total, transaction) => total + transaction.amount, 0);
+        setIncome(income);
+    }
+
+    const mergeTransactions = (transactions: Transaction[]): Transaction[] => {
+        const mergedTransactions: { [nameTo: string]: Transaction } = {};
+        const paymentTransactions = transactions.filter(transaction => transaction.TransactionType === TransactionType.PAYMENT);
+        let numberUniqueTransactions = 0;
+        let flag = true;
+
+        paymentTransactions.forEach(transaction => {
+            const key = transaction.nameTo;
+            if (!mergedTransactions[key] && numberUniqueTransactions < THRESHOLD_UNIQUE_TRANSACTIONS) {
+                mergedTransactions[key] = {...transaction};
+                numberUniqueTransactions++;
+            } else if (!mergedTransactions[key] && flag) {
+                mergedTransactions[OTHER_TRANSACTIONS] = {...transaction, nameTo: OTHER_TRANSACTIONS};
+                flag = false;
+            } else if (!mergedTransactions[key]) {
+                mergedTransactions[OTHER_TRANSACTIONS].amount += transaction.amount;
+            } else {
+                mergedTransactions[key].amount += transaction.amount;
+            }
+        });
```

Comment:
In short, this code is bad. For example, `!mergedTransactions[key] && flag` what does it mean?

I'm not going to try to figure it out . I assume you calculate here data for your chart. This is not task of the client app. 

For example, you have to show this chart on your web app with this beautiful code. Then you create mobile app and you have to duplicate code of calculating it. Moreover, from backend you get thousands of transactions and some phones or even old laptops are slow to calculate it fast. But this is a home page, most commonly used.

Your backend must always do computation and provide for frontend prepared data. In general, your backend servers more powerful than average user, you can cache data to increase speed, you don't need duplicate code and you decrease amount of data you send over the network. In our case, instead of sending thousands of transactions for the whole year, you will send only 4 types of expenses and one with income; 5 vs thousands

Reply:
so i will remove this method and move it to the backend in [TSK-291] ?

Reply:
Yes, but you have to do it in mongo aggregation